### PR TITLE
Mongo Java Driver 3.0 Upgrade

### PR DESCRIPTION
**Pull request by BVulaj**
Do not merge this yet, but please review.  Tests are still failing due to some underlying changes that I'm still looking into.  

Compilation errors due to the major version upgrade have been fixed.

### CODE REVIEW

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocSaver.java [hunk 1]
```diff
@@ -193,16 +189,11 @@ private WriteResult insertDoc(CRUDOperationContext ctx,
             Set<Path> paths = roleEval.getInaccessibleFields_Insert(inputDoc);
             LOGGER.debug("Inaccessible fields:{}", paths);
             if (paths == null || paths.isEmpty()) {
-                try {
                     ctx.getFactory().getInterceptors().callInterceptors(InterceptPoint.PRE_CRUD_INSERT_DOC, ctx, inputDoc);
                     WriteResult r = new InsertCommand(collection, dbObject).executeAndUnwrap();
                     inputDoc.setCRUDOperationPerformed(CRUDOperation.INSERT);
                     ctx.getFactory().getInterceptors().callInterceptors(InterceptPoint.POST_CRUD_INSERT_DOC, ctx, inputDoc);
                     return r;
```
Comment (bserdar) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocSaver.java]: Why did you remove the duplicate key error here?

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoLocking.java [hunk 2]
```diff
@@ -80,12 +80,8 @@ private boolean acquire(String callerId,String resourceId, Long ttl, Date now, D
...
             append(VERSION,1);
 
-        try {
-            LOGGER.debug("insert: {}",update);
-            coll.insert(update,WriteConcern.SAFE);
-        } catch (MongoException.DuplicateKey e) {
-            return false;
-        }
+        LOGGER.debug("insert: {}", update);
+        coll.insert(update, WriteConcern.ACKNOWLEDGED);
```
Comment (bserdar) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoLocking.java]: Ack is not enough. Why remove dup exception?

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoLocking.java [hunk 3]
```diff
@@ -170,7 +166,7 @@ Creating an atomic acquire() method in mongodb is not
                     append("$inc",new BasicDBObject(VERSION,1).
                            append(COUNT,1));
                 LOGGER.debug("update: {} {}",query,update);
-                wr=coll.update(query,update,false,false,WriteConcern.SAFE);
+                wr=coll.update(query,update,false,false,WriteConcern.ACKNOWLEDGED);
```
Comment (bserdar) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoLocking.java]: Why ack?

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoLocking.java (hunk 2 above)
Comment (BVulaj) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoLocking.java]: Bad change on my part.  Should have substituted for DuplicateKeyException.  Fixing.

Comment (BVulaj) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoLocking.java]: SAFE is deprecated for ACK.  I can change back if needed.

http://api.mongodb.org/java/current/com/mongodb/WriteConcern.html#SAFE

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoLocking.java (hunk 3 above)
Comment (BVulaj) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoLocking.java]: Deprecated. http://api.mongodb.org/java/current/com/mongodb/WriteConcern.html#SAFE

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocSaver.java (hunk 1 above)
Comment (BVulaj) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocSaver.java]: Bad change on my part.  Should have just been swapped for DuplicateKeyException.  Changing.

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoLocking.java (hunk 2 above)
Comment (bserdar) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoLocking.java]: ACK should be ok, then.

**Comment by BVulaj** [comment]
Internal regressions solved.  Tests passing.

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoLocking.java [hunk 4]
```diff
@@ -58,10 +57,11 @@ public void init(DBCollection coll) {
         this.coll=coll;
         BasicDBObject keys=new BasicDBObject(RESOURCEID,1);
         BasicDBObject options=new BasicDBObject("unique",1);
-        this.coll.ensureIndex(keys,options);
+        // ensureIndex was deprecated, changed to an alias of createIndex, and removed in a more recent version
+        this.coll.createIndex(keys,options);
     }
 
-    public void setDefaultTTL(long l) {
+    public void acquire(long l) {
```
Comment (bserdar) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoLocking.java]: ??????????

File: mongo/src/main/java/com/redhat/lightblue/mongo/metadata/MongoMetadata.java [hunk 5]
```diff
@@ -283,30 +283,20 @@ public void createNewMetadata(EntityMetadata md) {
                 try {
                     WriteResult result = new InsertCommand(collection, infoObj).executeAndUnwrap();
                     LOGGER.debug("Inserted entityInfo");
-                    String error = result.getError();
-                    if (error != null) {
-                        LOGGER.error("createNewMetadata: error in createInfo: {}" + error);
-                        throw Error.get(MongoMetadataConstants.ERR_DB_ERROR, error);
-                    }
-                } catch (MongoException.DuplicateKey dke) {
+
```
Comment (bserdar) [inline mongo/src/main/java/com/redhat/lightblue/mongo/metadata/MongoMetadata.java]: Why did you remove the error check? Was it useless?

File: mongo/src/main/java/com/redhat/lightblue/mongo/metadata/MongoMetadata.java [hunk 6]
```diff
@@ -446,23 +436,19 @@ public void createNewSchema(EntityMetadata md) {
...
                     LOGGER.debug("Rewriting {}",md.getEntitySchema().getVersion().getValue());
                     result = new SaveCommand(collection,schemaObj).executeAndUnwrap();
                 } else
                     throw dke;
             }
-            String error = result.getError();
-            if (error != null) {
-                throw Error.get(MongoMetadataConstants.ERR_DB_ERROR, error);
-            }
             if (listener != null) {
```
Comment (bserdar) [inline mongo/src/main/java/com/redhat/lightblue/mongo/metadata/MongoMetadata.java]: Why remove the error check?

File: mongo/src/main/java/com/redhat/lightblue/mongo/metadata/MongoMetadata.java [hunk 7]
```diff
@@ -530,10 +516,6 @@ public void setMetadataStatus(String entityName,
 
             query = new BasicDBObject(LITERAL_ID, md.get(LITERAL_ID));
             WriteResult result = new UpdateCommand(collection, query, (DBObject) mdParser.convert(schema), false, false).executeAndUnwrap();
-            String error = result.getError();
-            if (error != null) {
-                throw Error.get(MongoMetadataConstants.ERR_DB_ERROR, error);
-            }
             if(cache!=null)
```
Comment (bserdar) [inline mongo/src/main/java/com/redhat/lightblue/mongo/metadata/MongoMetadata.java]: Why remove the error check?

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoLocking.java (hunk 4 above)
Comment (BVulaj) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoLocking.java]: Let me look into this.  I don't remember changing this at all.

File: mongo/src/main/java/com/redhat/lightblue/mongo/metadata/MongoMetadata.java (hunk 5 above)
Comment (BVulaj) [inline mongo/src/main/java/com/redhat/lightblue/mongo/metadata/MongoMetadata.java]: Both deprecated (and now removed) and useless.  Error message will be in the exception thrown for an unsuccessful write operation.

http://api.mongodb.org/java/2.14/com/mongodb/WriteResult.html#getError()
